data/model/namespacequota.py

```diff
@@ -12,305 +12,210 @@
-                "bytes_allowed": config.app_config.get("DEFAULT_SYSTEM_REJECT_QUOTA_BYTES") * 0.8,
-                "type_id": 1,
-            },
-            {
-                "limit_bytes": config.app_config.get("DEFAULT_SYSTEM_REJECT_QUOTA_BYTES"),
-                "percent_of_limit": 100,
-                "name": "System Reject Limit",
-                "bytes_allowed": config.app_config.get("DEFAULT_SYSTEM_REJECT_QUOTA_BYTES"),
-                "type_id": 2,
-            },
-        ]
+        raise InvalidNamespaceQuota("Invalid quota size limit value: '%s'" % limit_bytes)
 
-    return query
 
+def update_namespace_quota_size(quota, limit_bytes):
+    if limit_bytes >= 0:
+        quota.limit_bytes = limit_bytes
+        quota.save()
+    else:
+        raise InvalidNamespaceQuota("Invalid quota size limit value: '%s'" % limit_bytes)
 
-def get_namespace_limit(name, quota_type_id, percent_of_limit):
-    try:
-        quota = get_namespace_quota(name)
 
-        if quota is None:
-            raise InvalidUsernameException("Quota for this namespace does not exist")
+def delete_namespace_quota(quota):
+    with db_transaction():
+        limits = QuotaLimits.delete().where(QuotaLimits.quota == quota).execute()
+        quota.delete_instance()
 
-        query = (
-            QuotaLimits.select()
-            .join(QuotaType)
-            .where(QuotaLimits.quota_id == quota.id)
-            .where(QuotaLimits.quota_type_id == quota_type_id)
-            .where(QuotaLimits.percent_of_limit == percent_of_limit)
-        )
 
-        return query.get()
+def _quota_type(type_name):
+    if type_name.lower() == "warning":
+        return QuotaTypes.WARNING
+    elif type_name.lower() == "reject":
+        return QuotaTypes.REJECT
+    raise InvalidNamespaceQuotaType("Quota type must be one of ['warning', 'reject']")
 
-    except QuotaLimits.DoesNotExist:
-        return None
 
+def get_namespace_quota_limit_list(quota, quota_type=None, percent_of_limit=None):
+    if percent_of_limit and (not percent_of_limit >= 0 or not percent_of_limit <= 100):
+        raise InvalidNamespaceQuotaLimit("Quota limit threshold must be between 1 and 100")
 
-def get_namespace_limit_from_id(name, quota_limit_id):
-    try:
-        quota = get_namespace_quota(name)
+    query = QuotaLimits.select().where(QuotaLimits.quota == quota)
 
-        if quota is None:
-            raise InvalidUsernameException("Quota for this namespace does not exist")
+    if quota_type:
+        quota_type_name = _quota_type(quota_type)
+        quota_type_ref = QuotaType.get(name=quota_type_name)
+        query = query.where(QuotaLimits.quota_type == quota_type_ref)
 
-        query = (
-            QuotaLimits.select()
-            .join(QuotaType)
-            .where(QuotaLimits.quota_id == quota.id)
-            .where(QuotaLimits.id == quota_limit_id)
-        )
+    if percent_of_limit:
+        query = query.where(QuotaLimits.percent_of_limit == percent_of_limit)
```

Comment:
The quota type reference is needed to create the record. The db id could be hardcoded, but that's bot very reliable.
It's a pretty small table though, with less than 10 records, and likely not to grow.
The main query itself shouldn't be getting executed until `list()` is applied to the query object.

Reply:
yes but it still going to cost at lest 200-400ms to go to the db for that record and based on the numbers that can result in roughly 5000hrs of extra CPU time per day at our rate of requests. You can instead use a join to create the quota limit since you already have the name which would remove that second request to get the quota type id.

Reply:
Made the quota type as part of a subquery.